### ISSUE TITLE
EgmClock message does not represent UNIX time

### ISSUE THREAD

**Issue by mkatliar** [ros-industrial/abb_libegm#119]
The definition of `EgmClock` in [proto/egm.proto](https://github.com/ros-industrial/abb_libegm/blob/e0b8c3b0a906a50a7ad8d1d1b465a4343369c4e8/proto/egm.proto) is 

```protobuf
message EgmClock				// Time in seconds and microseconds since 1 Jan 1970
{
    required uint64 sec = 1;
    required uint64 usec = 2;
}
```
Below are headers of EGM messages sent from **RobotStudio** on my system:
```
header {
  seqno: 352216
  tm: 28397221
  mtype: MSGTYPE_DATA
}
feedBack {
  joints {
    joints: 0.00025089801056310534
    joints: -0.010418600402772427
    joints: 0.03861050307750702
    joints: -0.0001342049945378676
    joints: 45.01199722290039
    joints: -0.0667094737291336
  }
  cartesian {
    pos {
      x: 1216.9378662109375
      y: 0.0046911695972085
      z: 1187.1817626953125
    }
    orient {
      u0: 0.3823593258857727
      u1: -0.0005403859540820122
      u2: 0.9240135550498962
      u3: -0.00022283525322563946
    }
    euler {
      x: -179.93319209849528
      y: 44.95977143802771
      z: -179.90533858705996
    }
  }
  time {
    sec: 28397
    usec: 216500
  }
}
...
```
The time data `sec: 28397` does not look like "time in seconds and microseconds since 1 Jan 1970".

**Comment by gavanderhoorn** [comment]
So just to be clear: this repository is *not* where ABB maintains its EGM implementation.

This is an OSS library which implements an interface to EGM running on an ABB (robot) controller.

Issues with EGM itself -- as it seems yours is -- should be either reported to your local ABB branch office (which will then escalate them if/when appropriate) or posted on one of the official ABB fora.

From your description, I don't have the impression the issue you are reporting is specific to this library.

If you agree, please close the issue.

If you post on the RobotStudio forum fi, it would be appreciated if you could post a final comment here with a link to your post there, so we can keep things connected.

**Comment by gavanderhoorn** [comment]
Having written that: I could imagine that RobotStudio uses a non-wallclock internally for the robot controller, and the timestamps you show are relative to when the virtual controller was started.

For `28397` specifically, that would be about 8 hours ago.

**Comment by mkatliar** [comment]
@gavanderhoorn I agree. I thought maybe the documentation needs to be changed. What is the source of information about what `EgmClock` should contain? I could not find it in EGM documentaion.

**Comment by gavanderhoorn** [comment]
Perhaps @jontje can clarify where `EgmClock` comments come from.

They were added in #32, which copied the files distributed with RobotStudio.

**Comment by mkatliar** [comment]
@gavanderhoorn what is the upstream?

**Comment by gavanderhoorn** [comment]
The EGM developers @ ABB Robotics.

**Comment by jontje** [comment]
> Perhaps @jontje can clarify where `EgmClock` comments come from.
> 
> They were added in #32, which copied the files distributed with RobotStudio.

If I remember correctly, then `EgmClock` was added to the EGM protocol in RobotWare `6.07`, but the release notes don’t mention it at all as far as I can see.

**Comment by gavanderhoorn** [comment]
@mkatliar: would it be OK to close this issue?

Have you reached out to ABB on the RobotStudio fora?

We could potentially add a clarifying comment here, but if it's not necessary that would be great.

**Comment by gavanderhoorn** [comment]
Closing.

**Comment by mkatliar** [comment]
The issue is not present on the real robot, so it looks like a RobotStudio bug. I started a discussion on the ABB forum: https://forums.robotstudio.com/discussion/12743

**Comment by mkatliar** [comment]
This is the response that I got from ABB support:
```
according to our support EGM is not supported by RS this is why You notice different behavior of EGM clock. 

"
I checked with RobotStudio R&D and they advised that RobotStudio doesn't do anything with EGM, so it is not a bug with RobotStudio, instead it is not designed to work with it.
"
```

**Comment by gavanderhoorn** [comment]
That's a very interesting response, as both @jontje and me use RobotStudio to test and work with EGM.

@jontje ?

**Comment by mkatliar** [comment]
@gavanderhoorn I was also surprised. If it "doesn't do anything with EGM", why does it support EGM commands and even send EGM data.

**Comment by gavanderhoorn** [comment]
Perhaps a lost in translation? Perhaps "doesn't do anything with EGM" actually means: "RobotStudio does not influence any of the internals of EGM".

As I wrote in https://github.com/ros-industrial/abb_libegm/issues/119#issuecomment-786703538:

> I could imagine that RobotStudio uses a non-wallclock internally for the virtual robot controller, and the timestamps you show are relative to when the virtual controller was started.
> 
> For `28397` specifically, that would be about 8 hours ago.

perhaps that's what's meant by "RS doesn't do anything with EGM", as in: RS uses *some* internal clock, doesn't adjust anything in EGM so EGM just reports whatever clock RS happens to use.

But this is all (wild) speculation by me (and I don't work for ABB).

**Comment by mkatliar** [comment]
"EGM is not supported by RS" also sounds interesting.

**Comment by gavanderhoorn** [comment]
Yes, I agree.

That's the first time I hear that stated about RS.

But then again, I'm just "a user".

**Comment by jontje** [comment]
> This is the response that I got from ABB support:
> 
> ```
> according to our support EGM is not supported by RS this is why You notice different behavior of EGM clock. 
> 
> "
> I checked with RobotStudio R&D and they advised that RobotStudio doesn't do anything with EGM, so it is not a bug with RobotStudio, instead it is not designed to work with it.
> "
> ```

I think there has been some misunderstanding somewhere in the support chain, and that response is very vague.

When you simulate a robot "in" RobotStudio, is is technically a background process that emulates RobotWare. And it is the RobotWare process that, for example, interpret RAPID code and manages EGM clients and communication.

But this is not something users generally don't know about, so it is reasonable to send bug reports etc. via the RobotStudio channels.

So the question should have been passed along to the RobotWare R&D, and not RobotStudio R&D.

**Comment by gavanderhoorn** [comment]
> So the question should have been passed along (by the support people) to the RobotWare R&D, and not RobotStudio R&D

yeah, that's in summary what I was trying to say in https://github.com/ros-industrial/abb_libegm/issues/119#issuecomment-800401636.

@mkatliar: you could perhaps rephrase your question to ABB support and see whether it ends up at the right people?

**Comment by mkatliar** [comment]
@gavanderhoorn I will.

**Comment by mkatliar** [comment]
This is what I got from ABB support:
```
Regarding EgmClock : I don’t know. But usually we don’t care on that. The time from the first sequence is the reference time “Time 0”, then time difference from next sequence is calculated.
```

**Comment by gavanderhoorn** [comment]
Seems to confirm what I suspected in https://github.com/ros-industrial/abb_libegm/issues/119#issuecomment-786703538.

**Comment by mkatliar** [comment]
It was in a private email.

**Comment by mkatliar** [comment]
They agreed though that this is not consistent with the documentation:
```
Anyway I agree regarding documentation.
```

**Comment by gavanderhoorn** [comment]
Was there any indication the documentation / comments will be updated?

**Comment by mkatliar** [comment]
> Was there any indication the documentation / comments will be updated?

No

**Comment by mkatliar** [comment]
The issue seems to be fixed in RobotStudio 21.2.9526.0 -- the timestamps look like UNIX time now.

**Comment by gavanderhoorn** [comment]
Nice, so even if they stated this wouldn't be fixed, it has now been fixed?
